File: apps/ios-crux/PickCrux/ChatList.swift
import SwiftUI
import PickShared

/// Scrolling message list + live running-tools strip + conversation doc strip.
struct ChatList: View {
    @ObservedObject var core: CoreBridge

    var body: some View {
        VStack(spacing: 0) {
            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    ForEach(Array(core.vm.messages.enumerated()), id: \.offset) { _, msg in
                        MessageRow(message: msg)
                    }

                    // Live "running tools" strip while scanning.
                    if core.vm.scanInProgress && !core.vm.toolCalls.isEmpty {
                        ForEach(Array(core.vm.toolCalls.enumerated()), id: \.offset) { _, tool in
                            ToolCallRow(tool: tool)
                        }
                    }

                    if core.vm.scanInProgress && core.vm.toolCalls.isEmpty {
                        HStack(spacing: 8) {
                            ProgressView().tint(Theme.brand)
                            Text("Working...")
                                .font(.system(size: 14))
                                .foregroundStyle(Theme.muted)
                        }
                    }
                }
                .padding(16)
            }

            // Conversation-scoped documents strip pinned above the input.
            if !core.vm.conversationDocs.isEmpty {
                conversationDocsStrip
            }

            InputRow(core: core)
        }
    }

    private var conversationDocsStrip: some View {
        ScrollView {
            VStack(spacing: 0) {
                ForEach(Array(core.vm.conversationDocs.enumerated()), id: \.offset) { _, doc in
                    Button {
                        core.send(.openDocument(doc.id))
                    } label: {
                        HStack(spacing: 8) {
                            Image(systemName: "doc.text")
                                .font(.system(size: 14))
                                .foregroundStyle(Theme.muted)
                            Text(doc.title)
                                .font(.system(size: 14))
                                .foregroundStyle(Theme.text)
                            Spacer()
                        }
                        .padding(.vertical, 10)
                        .padding(.horizontal, 16)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .frame(maxHeight: 220)
        .background(Theme.faintWash)
        .overlay(Rectangle().frame(height: 1).foregroundStyle(Theme.hairline), alignment: .top)
    }
}

File: apps/android-crux/app/src/main/kotlin/com/strike48/pickcrux/ui/ChatList.kt
package com.strike48.pickcrux.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.strike48.pick.shared.DocRef
import com.strike48.pick.shared.MessageView
import com.strike48.pick.shared.ToolCallView

/** Scrolling message list with a live running-tools strip while scanning. */
@Composable
fun ChatList(
    messages: List<MessageView>,
    toolCalls: List<ToolCallView>,
    modifier: Modifier = Modifier,
) {
    LazyColumn(
        modifier = modifier.fillMaxWidth(),
        contentPadding = PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        items(messages) { msg -> MessageRow(msg) }
        if (toolCalls.isNotEmpty()) {
            item {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    Text(
                        text = "Running tools",
                        color = PickColors.Muted,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                    )
                    toolCalls.forEach { ToolCallRow(it) }
                }
            }
        }
    }
}

/** Conversation-scoped document strip pinned above the input row. */
@Composable
fun ConversationDocStrip(docs: List<DocRef>, onOpen: (String) -> Unit) {
    if (docs.isEmpty()) return
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .heightIn(max = 220.dp)
            .background(PickColors.FaintWash)
            .border(1.dp, PickColors.Hairline)
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 16.dp, vertical = 8.dp),
        verticalArrangement = Arrangement.spacedBy(2.dp),
    ) {
        Text(
            text = "Documents",
            color = PickColors.Muted,
            fontSize = 12.sp,
            fontWeight = FontWeight.Medium,
            modifier = Modifier.padding(bottom = 4.dp),
        )
        docs.forEach { doc ->
            Text(
                text = doc.title,
                color = PickColors.Text,
                fontSize = 14.sp,
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable { onOpen(doc.id) }
                    .padding(vertical = 8.dp),
            )
        }
    }
}
